### PR TITLE
refactor(sns): move subscribers to `aws-sns-subscribers`

### PR DESCRIPTION
**Pull request by rix0rrr**
In accordance with new guidelines, we're centralizing cross-service
integrations into their own package. In this case, centralizing
SNS Subscribers into @aws-cdk/aws-sns-subscribers.

BREAKING CHANGE: using a queue, lambda, email, URL as SNS Subscriber now
requires an integration object from the `@aws-cdk/aws-sns-subscribers`
package.

----

### Pull Request Checklist

* [ ] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ ] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [ ] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-codepipeline-actions/lib/manual-approval-action.ts [hunk 1]
```diff
@@ -62,7 +63,7 @@ export class ManualApprovalAction extends codepipeline.Action {
     if (this._notificationTopic) {
       this._notificationTopic.grantPublish(info.role);
       for (const notifyEmail of this.props.notifyEmails || []) {
-        this._notificationTopic.subscribeEmail(`Subscription-${notifyEmail}`, notifyEmail);
+        this._notificationTopic.subscribe(new subs.EmailSubscriber(notifyEmail));
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline-actions/lib/manual-approval-action.ts]: Rename to addSubscriber

File: packages/@aws-cdk/aws-sns/package.json [hunk 2]
```diff
@@ -77,25 +77,22 @@
...
     "@aws-cdk/cdk": "^0.33.0"
   },
   "engines": {
     "node": ">= 8.10.0"
   },
   "awslint": {
     "exclude": [
-      "construct-base-is-private:@aws-cdk/aws-sns.TopicBase"
+      "construct-base-is-private:@aws-cdk/aws-sns.TopicBase",
+      "integ-return-type:@aws-cdk/aws-sns.ITopicSubscription.bind"
```
Comment (eladb) [inline packages/@aws-cdk/aws-sns/package.json]: I think we should update the awslint rule to allow void as a return type for `bind`

File: packages/@aws-cdk/cdk/lib/construct.ts [hunk 3]
```diff
@@ -530,7 +530,7 @@ export class Construct implements IConstruct {
    * Return whether the given object is a Construct
    */
   public static isConstruct(x: any): x is Construct {
-    return CONSTRUCT_SYMBOL in x;
+    return typeof x === 'object' && x !== null && CONSTRUCT_SYMBOL in x;
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/construct.ts]: ouch!

File: packages/aws-cdk/lib/init-templates/app/csharp/src/HelloCdk/HelloStack.cs [hunk 4]
```diff
@@ -20,7 +21,7 @@ public HelloStack(Construct parent, string id, IStackProps props) : base(parent,
                 DisplayName = "My First Topic Yeah"
             });
 
-            topic.SubscribeQueue(queue, null);
+            topic.AddSubscription(new SqsSubscription(queue, null));
```
Comment (eladb) [inline packages/aws-cdk/lib/init-templates/app/csharp/src/HelloCdk/HelloStack.cs]: what's this `null`?

File: packages/@aws-cdk/cdk/lib/construct.ts (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/cdk/lib/construct.ts]: We should make sure all of our `isXxx` methods are implemented like this.

File: packages/aws-cdk/lib/init-templates/app/csharp/src/HelloCdk/HelloStack.cs (hunk 4 above)
Comment (rix0rrr) [reply packages/aws-cdk/lib/init-templates/app/csharp/src/HelloCdk/HelloStack.cs]: I believe .NET does not support optional arguments

**Pull request merged**